include/crm/common/nvpair.h

```diff
@@ -23,16 +23,72 @@ extern "C" {
 #  include <libxml/tree.h>  // xmlNode
 #  include <crm/crm.h>
 
-typedef struct pcmk_nvpair_s {
-    char *name;
-    char *value;
-} pcmk_nvpair_t;
-
-GList *pcmk_prepend_nvpair(GList *nvpairs, const char *name, const char *value);
-void pcmk_free_nvpairs(GList *nvpairs);
-GList *pcmk_sort_nvpairs(GList *list);
-GList *pcmk_xml_attrs2nvpairs(xmlNode *xml);
-void pcmk_nvpairs2xml_attrs(GList *list, xmlNode *xml);
+/*
+ *  nvpairs (list of nvpair objects) encapsulating datatype
+ *  + basic public operations
+ */
+
+/* name that wouldn't need to be exposed if -fms-extensions CFLAG was a norm */
+#define pcmk_nvpairs_s _GSList
+struct pcmk_nvpairs_s;
+
+/*!
+ * \brief Opaque encapsulation of a list of name/value pairs
+ */
+typedef struct pcmk_nvpairs_s pcmk_nvpairs_t;
```

Comment:
This is the type of complexity we want to avoid at all costs. Simplicity is a virtue:

    typedef GSList pcmk_nvpairs_t;

would be a superior choice. We don't want contributors to have to care about compiler options, we don't want to introduce dependencies on internal symbols of external libraries, and we don't want to introduce multiple levels of indirection that change a two-second glance into a reading and substitution exercise.

However, in this case, it was a conscious choice to expose pcmk_nvpair_t and GList, so it was obvious to callers that pcmk_nvpair_t could be used in other collection types as appropriate, and that standard glib GList functions could be used with the list result.

Reply:
> This is the type of complexity we want to avoid at all costs.
>
> [...] it was a conscious choice to expose pcmk_nvpair_t and GList,
> so it was obvious to callers that pcmk_nvpair_t could be used in
> other collection types as appropriate, and that standard glib GList
> functions could be used with the list result.

Well, you've confirmed me in the context of this change that it's
a good thing to have nvpair as an opaque data type.

Otherwise I can't explain why you've tried so hard in 4609d60
in making the con-/de-structor private for `pcmk_nvpair_t`, and
also other new functions suggest you wanted to abstract dealing
with `pcmk_nvpair_t` away from its internal layout.

I think you contradict yourself here, jumping between two opposite
approaches.

Anyway, `pcmk_nvpairs_t` type I introduced had the same abstracting
gist, and I think that if this approach was exercised properly
throughout the pacemaker codebase, it would then be trivial to flip
particular underlying container type (list, hashtable) easily as long
as the necessary operations were implemented, incl. switching libraries
implementing these types -- all without externally observable ABI
breakage as long as the same library version is used throughout the
data type instance life-cycle.

It is also a good practice to hide these internals unless they are
committed as stable, so that the client code can/shall make no
assumptions about the particular layouts/implementations, since all
it get to work with are pointers to opaque data types plus public
manipulators.  It'd be good if we can agree on this as well.

> we don't want to introduce multiple levels of indirection that
> change a two-second glance into a reading and substitution exercise

Depends on the usage, but in public header, the careful approach
I tried to explain and defend above should prevail unless something
is 100% committed for the time being.

And this why I arranged things like that for an entirely new data
type in the API, and then you come and break it all down without
much reasoning.  That's not the collaboration I'd want to see,
you know, throwing possibly reasonable work of others down,
leaving "I must know better" aftertaste of distrust behind.

Reply:
> Well, you've confirmed me in the context of this change that it's
> a good thing to have nvpair as an opaque data type.
> 
> Otherwise I can't explain why you've tried so hard in [4609d60](https://github.com/ClusterLabs/pacemaker/commit/4609d60b8d9f7c6fe3d9c724779f484d6e57127e)
> in making the con-/de-structor private for `pcmk_nvpair_t`, and
> also other new functions suggest you wanted to abstract dealing
> with `pcmk_nvpair_t` away from its internal layout.
> 
> I think you contradict yourself here, jumping between two opposite
> approaches.

To a degree, yes, but only because this is such a simple type.

Normally, I am all for making public types completely opaque, so we can change them without worrying about backward compatibility. But this is more of a convenience type, closer to "typedef int *intptr" in intent, than an abstract data type. It's intended to only ever be a name and a value. It's rather more comparable to GList itself, which exposes its three members (data/next/prev), quite unlike truly abstract types like GHashTable.

To put it another way, pcmk_nvpair_t is intended just as a box to put a name and value in, and the public functions are merely convenience functions rather than data type methods. The constructor/destructor aren't intentionally hidden, they're simply static by virtue of only being needed by the convenience functions in the one file. 

> Anyway, `pcmk_nvpairs_t` type I introduced had the same abstracting
> gist, and I think that if this approach was exercised properly
> throughout the pacemaker codebase, it would then be trivial to flip
> particular underlying container type (list, hashtable) easily as long
> as the necessary operations were implemented, incl. switching libraries
> implementing these types -- all without externally observable ABI
> breakage as long as the same library version is used throughout the
> data type instance life-cycle.

I'm completely in agreement for any complex type. The vast majority of our existing types really should be opaque. But I do think there are worthwhile exceptions for simple, immutable-in-practice types (another example might be lrmd_rsc_info_t) as well as collections of function pointers used as a C method of deriving types (e.g. cib_api_operations_t). One could make an argument for not having these exceptions, but I don't see a compelling benefit.
 
> It is also a good practice to hide these internals unless they are
> committed as stable, so that the client code can/shall make no
> assumptions about the particular layouts/implementations, since all
> it get to work with are pointers to opaque data types plus public
> manipulators. It'd be good if we can agree on this as well.
> 
> > we don't want to introduce multiple levels of indirection that
> > change a two-second glance into a reading and substitution exercise
> 
> Depends on the usage, but in public header, the careful approach
> I tried to explain and defend above should prevail unless something
> is 100% committed for the time being.
> 
> And this why I arranged things like that for an entirely new data
> type in the API, and then you come and break it all down without
> much reasoning. That's not the collaboration I'd want to see,
> you know, throwing possibly reasonable work of others down,
> leaving "I must know better" aftertaste of distrust behind.

I suppose you missed the irony here; this entire PR was an "I must know better" exercise on your part. You didn't like my design or style, and took it upon yourself to redo it, despite it being correct code. That said, I do agree that GSList is a better choice here, so I've done a simple substitution in the checkout I'm preparing for the next release candidate.